Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp

```diff
@@ -416,29 +416,58 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
+        Array4<Real> const& mag_Ms_arr = mag_Ms_mf.array(mfi);
+
+        // mu_mf will be imported but will only be called at grids where Ms == 0
+        auto& mu_mf = macroscopic_properties->getmu_mf();
+        Array4<Real> const& mu_arr = mu_mf.array(mfi);
+
+        amrex::Real const mu0_inv = 1. / PhysConst::mu0;
 
         // Loop over the cells and update the fields
         amrex::ParallelFor(tbx, tby, tbz,
             [=] AMREX_GPU_DEVICE(int i, int j, int k) {
-                Hx(i, j, k) += mu0_inv * dt * T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k)
-                             - mu0_inv * dt * T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k);
-                if (coupling == 1) {
-                    Hx(i, j, k) += - M_xface(i, j, k, 0) + M_xface_old(i, j, k, 0);
+                Real mag_Ms_arrx = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(1,0,0), mag_Ms_arr);
+                if (mag_Ms_arrx == 0._rt){ // nonmagnetic region
+                    Real mu_arrx = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(1,0,0), mu_arr);
+                    Hx(i, j, k) += 1. / mu_arrx * dt * T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k)
+                                 - 1. / mu_arrx * dt * T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k);
+                } else if (mag_Ms_arrx > 0){ // magnetic region
+                    Hx(i, j, k) += mu0_inv * dt * T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k)
+                                 - mu0_inv * dt * T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k);
+                    if (coupling == 1) {
+                        Hx(i, j, k) += - M_xface(i, j, k, 0) + M_xface_old(i, j, k, 0);
+                    }
                 }
             },
             [=] AMREX_GPU_DEVICE(int i, int j, int k) {
-                Hy(i, j, k) += mu0_inv * dt * T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k)
-                             - mu0_inv * dt * T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k);
-                if (coupling == 1) {
-                    Hy(i, j, k) +=  - M_yface(i, j, k, 1) + M_yface_old(i, j, k, 1);
+                Real mag_Ms_arry = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(0,1,0), mag_Ms_arr);
+                if (mag_Ms_arry == 0._rt){ // nonmagnetic region
+                    Real mu_arry = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(0,1,0), mu_arr);
+                    Hy(i, j, k) += 1. / mu_arry * dt * T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k)
+                                 - 1. / mu_arry * dt * T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k);
+                } else if (mag_Ms_arry > 0){ // magnetic region
+                    Hy(i, j, k) += mu0_inv * dt * T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k)
+                                 - mu0_inv * dt * T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k);
```

Comment:
```suggestion
                    Hy(i, j, k) += mu0_inv * dt * (T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k)
                                                         - T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k));
```

Reply:
somehow in my suggestion the indentation does not look clean. Can you check this on your terminal? Thank you